src/librustc_trans/abi.rs

```diff
@@ -317,6 +318,14 @@ impl FnType {
                 if ty.is_integral() {
                     arg.signedness = Some(ty.is_signed());
                 }
+                // Rust enum types that map onto C enums (LLVM integers) also
+                // need to follow the target ABI zero-/sign-extension rules.
+                if let ty::TyEnum(..) = ty.sty {
+                    if arg.ty.kind() == llvm::Integer {
+                        let repr = adt::represent_type(ccx, ty);
+                        arg.signedness = Some(adt::is_discr_signed(&repr));
```

Comment:
This should pattern-match on `ccx.layout_of(ty)` (see `librustc/ty/layout.rs`) instead of checking the LLVM type and then asking `adt` (which is being phased out by #36151) for more information.